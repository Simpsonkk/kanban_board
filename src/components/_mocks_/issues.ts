import { Issue } from '../../types/issue.types';

export const mockIssues: Issue[] = [
  {
    id: 444,
    node_id: 'string',
    url: 'string',
    repository_url: 'string',
    labels_url: 'string',
    comments_url: 'string',
    events_url: 'string',
    html_url: 'string',
    number: 9,
    state: 'string',
    title: 'my issue',
    body: 'string',
    user: {
      login: 'simpson',
      id: 3,
      node_id: 'string',
      avatar_url: 'string',
      gravatar_id: 'string',
      url: 'string',
      html_url: 'string',
      followers_url: 'string',
      following_url: 'string',
      gists_url: 'string',
      starred_url: 'string',
      subscriptions_url: 'string',
      organizations_url: 'string',
      repos_url: 'string',
      events_url: 'string',
      received_events_url: 'string',
      type: 'string',
      site_admin: true,
    },
    labels: [
      {
        id: 3,
        node_id: 'string',
        url: 'string',
        name: 'string',
        description: 'string',
        color: 'string',
        default: true,
      },
    ],
    assignee: {
      login: 'Simpson',
      id: 4,
      node_id: 'string',
      avatar_url: 'string',
      gravatar_id: 'string',
      url: 'string',
      html_url: 'string',
      followers_url: 'string',
      following_url: 'string',
      gists_url: 'string',
      starred_url: 'string',
      subscriptions_url: 'string',
      organizations_url: 'string',
      repos_url: 'string',
      events_url: 'string',
      received_events_url: 'string',
      type: 'string',
      site_admin: true,
    },
    assignees: [
      {
        login: 'Sssimpson',
        id: 4,
        node_id: 'string',
        avatar_url: 'string',
        gravatar_id: 'string',
        url: 'string',
        html_url: 'string',
        followers_url: 'string',
        following_url: 'string',
        gists_url: 'string',
        starred_url: 'string',
        subscriptions_url: 'string',
        organizations_url: 'string',
        repos_url: 'string',
        events_url: 'string',
        received_events_url: 'string',
        type: 'string',
        site_admin: true,
      },
    ],
    milestone: {
      url: 'string',
      html_url: 'string',
      labels_url: 'string',
      id: 9,
      node_id: 'string',
      number: 99,
      state: 'string',
      title: 'issssue',
      description: 'string',
      creator: {
        login: 'sssimps',
        id: 99,
        node_id: 'string',
        avatar_url: 'string',
        gravatar_id: 'string',
        url: 'string',
        html_url: 'string',
        followers_url: 'string',
        following_url: 'string',
        gists_url: 'string',
        starred_url: 'string',
        subscriptions_url: 'string',
        organizations_url: 'string',
        repos_url: 'string',
        events_url: 'string',
        received_events_url: 'string',
        type: 'string',
        site_admin: true,
      },
      open_issues: 9,
      closed_issues: 999,
      created_at: 'string',
      updated_at: 'string',
      closed_at: 'string',
      due_on: 'string',
    },
    locked: false,
    active_lock_reason: null,
    comments: 99,
    pull_request: {
      url: 'string',
      html_url: 'string',
      diff_url: 'string',
      patch_url: 'string',
    },
    closed_by: null,
    closed_at: null,
    created_at: 'string',
    updated_at: 'string',
    author_association: 'string',
    state_reason: 'string',
  },
];
