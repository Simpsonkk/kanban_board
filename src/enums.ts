export enum AppRoute {
  Main = '/',
}

export enum APIRoute {
  Issues = 'issues',
  State = 'state',
  Page = 'page',
}
