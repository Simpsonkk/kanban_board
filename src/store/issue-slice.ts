import { createSlice, PayloadAction } from '@reduxjs/toolkit';

import { getIssueOrder } from '../services/issue-order';
import { Issue } from '../types/issue.type';
import { IssueState } from '../types/state.type';

const initialState: IssueState = {
  issueColumns: {
    1: {
      name: 'To do',
      items: [],
    },
    2: {
      name: 'In Progress',
      items: [],
    },
    3: {
      name: 'Done',
      items: [],
    },
  },
  nextIssuePage: '1',
  repositoryName: '',
};

export const issueSlice = createSlice({
  name: 'issueSlice',
  initialState,
  reducers: {
    loadIssues: (state, action: PayloadAction<Issue[]>) => {
      const todo: Issue[] = [];
      const inProgress: Issue[] = [];
      const done: Issue[] = [];

      // if (issue.id === 1669119666) {
          
      // }
      // let priority = 10000


      action.payload.forEach((issue) => {
        // issue[priority] = priority

        if (issue.state === 'open') {
          if (!issue.assignee) {
            todo.push(issue);
          } else {
            inProgress.push(issue);
          }
        } else if (issue.state === 'closed') {
          done.push(issue);
        }
        // priority += 10000
      });

      const issueOrder = JSON.parse(getIssueOrder())
      const currentRepo = issueOrder.find((repositories) => repositories.repositoryName === state.repositoryName).displacedIssues

      // currentRepo.forEach((issue) => {

      // })


      state.issueColumns[1].items.push(...todo);
      state.issueColumns[2].items.push(...inProgress);
      state.issueColumns[3].items.push(...done);
    },
    checkIssueNextPage: (state, action: PayloadAction<string>) => {
      const issuePageInfo = action.payload.split(', ').reduce((acc, part) => {
        const match = part.match(/<([^>]+)>;\s*rel="([^"]+)"/);
        if (match) {
          const url = match[1];
          const rel = match[2];
          acc[rel] = url;
        }
        return acc;
      }, {} as { [key: string]: string });

      state.nextIssuePage = issuePageInfo.next?.match(/page=(\d+)/)![1];
    },
    loadRepositoryName: (state, action: PayloadAction<string>) => {
      state.repositoryName = action.payload;
    },
  },
});

export const { loadIssues, checkIssueNextPage, loadRepositoryName } = issueSlice.actions;
