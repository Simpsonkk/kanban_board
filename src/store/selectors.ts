import { IssueColumns } from '../types/issue-columns.type';
import { State } from '../types/state.type';

export const getIssueColumns = (state: State): IssueColumns => state.issueColumns;
export const getNextIssuePage = (state: State): string => state.nextIssuePage;
export const getRepositoryName = (state: State): string => state.repositoryName;

// [
//   {
//     repoName: 'React',
//     displacedIssues: [
//       {
//         issueId: 123,
//         column: 'Done',
            // index: 2
//       },
//       {
//         issueId: 321,
//         column: 'To Do',
//       },
//     ]
//   },
//   {
//     repoName: 'Rozetka',
//     displacedIssues: [
//       {
//         issueId: 456,
//         column: 'Done',
//       },
//       {
//         issueId: 789,
//         column: 'To Do',
//       },
//     ]
//   },
// ]
