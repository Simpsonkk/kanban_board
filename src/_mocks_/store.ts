import { mockIssues } from './issues';

export const fakeStore = {
  issueColumns: {
    1: {
      name: 'To do',
      items: [],
    },
    2: {
      name: 'In Progress',
      items: [],
    },
    3: {
      name: 'Done',
      items: [],
    },
  },
  repositoryName: 'react/facebook',
  isIssuesLoaded: 'loaded',
};

export const fakeFullStore = {
  issueColumns: {
    1: {
      name: 'To do',
      items: mockIssues,
    },
    2: {
      name: 'In Progress',
      items: [],
    },
    3: {
      name: 'Done',
      items: [],
    },
  },
  repositoryName: 'react/facebook',
  isIssuesLoaded: 'loaded',
};
