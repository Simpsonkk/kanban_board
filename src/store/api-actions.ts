import { AxiosInstance } from 'axios';

import { createAsyncThunk } from '@reduxjs/toolkit';

import { APIRoute } from '../enums';
import { errorHandler } from '../services/error-handler';
import { IssueRequest } from '../types/issue-request.type';
import { AppDispatch } from '../types/state.type';
import { checkIssueNextPage, loadIssues } from './issue-slice';

export const fetchRepoIssues = createAsyncThunk<
  void,
  IssueRequest,
  { extra: { api: AxiosInstance }; dispatch: AppDispatch }
>('fetchRepoIssues', async ({ repositoryName, nextIssuePage }, { extra: { api }, dispatch }) => {
  try {
    const response = await api.get(
      `${repositoryName}/${APIRoute.Issues}?${APIRoute.State}=all&${APIRoute.Page}=${nextIssuePage}`
    );
    dispatch(loadIssues(response.data));
    dispatch(checkIssueNextPage(response.headers.link));
  } catch (error) {
    errorHandler(error);
  }
});
