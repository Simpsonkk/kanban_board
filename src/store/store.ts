import { configureStore } from '@reduxjs/toolkit';

import { createAPI } from '../services/api';
import { issueSlice } from './issue-slice';

const api = createAPI();

export const store = configureStore({
  reducer: issueSlice.reducer,
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware({
      thunk: {
        extraArgument: { api },
      },
    }),
});
