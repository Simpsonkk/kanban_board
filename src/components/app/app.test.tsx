import '@testing-library/jest-dom';

import React from 'react';
import { Provider } from 'react-redux';
import { MemoryRouter, Route, Routes } from 'react-router-dom';
import configureStore from 'redux-mock-store';

import { render } from '@testing-library/react';

import { fakeStore } from '../../_mocks_/store';
import { AppRoute } from '../../enums';
import KanbanBoardPage from '../../pages/kanban-board-page/kanban-board-page';

const mockStore = configureStore();

describe('KanbanBoardPage', () => {
  it('render KanbanBoardPage', () => {
    const store = mockStore(fakeStore);

    const view = render(
      <Provider store={store}>
        <MemoryRouter>
          <Routes>
            <Route path={AppRoute.Main} element={<KanbanBoardPage />} />
          </Routes>
        </MemoryRouter>
      </Provider>
    );
    expect(view).toMatchSnapshot();
  });
});
