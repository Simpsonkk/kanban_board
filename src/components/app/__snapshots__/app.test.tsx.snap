// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KanbanBoardPage render KanbanBoardPage 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="container"
      >
        <div
          class="row justify-content-center mt-2"
          data-testid="searchForm"
        >
          <div
            class="col-9"
          >
            <input
              class="me-2 form-control form-control-lg"
              placeholder="Enter repo URL"
            />
          </div>
          <button
            class="col-2 me-2 btn btn-secondary btn-sm"
          >
            Load Issues
          </button>
          <div
            class="d-flex col-11 ms-0 mt-2"
          >
            <a
              class="text-decoration-none"
              data-testid="Profile"
              href="https://github.com/"
              rel="noreferrer"
              target="_blank"
            >
              Profile
            </a>
            <p
              class="ms-2 me-2"
            >
              &gt;
            </p>
            <a
              class="text-decoration-none"
              data-testid="Repository"
              href="https://github.com/"
              rel="noreferrer"
              target="_blank"
            >
              Repository
            </a>
          </div>
        </div>
        <div
          class="d-flex justify-content-center row"
          data-testid="issueTable"
        >
          <div
            class="d-flex flex-column align-items-center col"
          >
            <h2>
              To do
            </h2>
            <div
              class="m-1"
            >
              <div
                class="p-1 rounded"
                data-rbd-droppable-context-id="0"
                data-rbd-droppable-id="1"
                data-testid="droppable"
                style="background: lightgrey; width: 350px; height: 500px; overflow-y: auto;"
              />
            </div>
          </div>
          <div
            class="d-flex flex-column align-items-center col"
          >
            <h2>
              In Progress
            </h2>
            <div
              class="m-1"
            >
              <div
                class="p-1 rounded"
                data-rbd-droppable-context-id="0"
                data-rbd-droppable-id="2"
                data-testid="droppable"
                style="background: lightgrey; width: 350px; height: 500px; overflow-y: auto;"
              />
            </div>
          </div>
          <div
            class="d-flex flex-column align-items-center col"
          >
            <h2>
              Done
            </h2>
            <div
              class="m-1"
            >
              <div
                class="p-1 rounded"
                data-rbd-droppable-context-id="0"
                data-rbd-droppable-id="3"
                data-testid="droppable"
                style="background: lightgrey; width: 350px; height: 500px; overflow-y: auto;"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      aria-atomic="true"
      aria-live="assertive"
      id="rbd-announcement-0"
      style="position: absolute; width: 1px; height: 1px; margin: -1px; border: 0px; padding: 0px; overflow: hidden; clip-path: inset(100%);"
    />
    <div
      id="rbd-hidden-text-0-hidden-text-0"
      style="display: none;"
    >
      
  Press space bar to start a drag.
  When dragging you can use the arrow keys to move the item around and escape to cancel.
  Some screen readers may require you to be in focus mode or to use your pass through key

    </div>
  </body>,
  "container": <div>
    <div
      class="container"
    >
      <div
        class="row justify-content-center mt-2"
        data-testid="searchForm"
      >
        <div
          class="col-9"
        >
          <input
            class="me-2 form-control form-control-lg"
            placeholder="Enter repo URL"
          />
        </div>
        <button
          class="col-2 me-2 btn btn-secondary btn-sm"
        >
          Load Issues
        </button>
        <div
          class="d-flex col-11 ms-0 mt-2"
        >
          <a
            class="text-decoration-none"
            data-testid="Profile"
            href="https://github.com/"
            rel="noreferrer"
            target="_blank"
          >
            Profile
          </a>
          <p
            class="ms-2 me-2"
          >
            &gt;
          </p>
          <a
            class="text-decoration-none"
            data-testid="Repository"
            href="https://github.com/"
            rel="noreferrer"
            target="_blank"
          >
            Repository
          </a>
        </div>
      </div>
      <div
        class="d-flex justify-content-center row"
        data-testid="issueTable"
      >
        <div
          class="d-flex flex-column align-items-center col"
        >
          <h2>
            To do
          </h2>
          <div
            class="m-1"
          >
            <div
              class="p-1 rounded"
              data-rbd-droppable-context-id="0"
              data-rbd-droppable-id="1"
              data-testid="droppable"
              style="background: lightgrey; width: 350px; height: 500px; overflow-y: auto;"
            />
          </div>
        </div>
        <div
          class="d-flex flex-column align-items-center col"
        >
          <h2>
            In Progress
          </h2>
          <div
            class="m-1"
          >
            <div
              class="p-1 rounded"
              data-rbd-droppable-context-id="0"
              data-rbd-droppable-id="2"
              data-testid="droppable"
              style="background: lightgrey; width: 350px; height: 500px; overflow-y: auto;"
            />
          </div>
        </div>
        <div
          class="d-flex flex-column align-items-center col"
        >
          <h2>
            Done
          </h2>
          <div
            class="m-1"
          >
            <div
              class="p-1 rounded"
              data-rbd-droppable-context-id="0"
              data-rbd-droppable-id="3"
              data-testid="droppable"
              style="background: lightgrey; width: 350px; height: 500px; overflow-y: auto;"
            />
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
